Added 'How to Secure API Keys Using NextJS Server Components' Cookbook (#587)
graphprotocol/docs: website/pages/en/cookbook/how-to-secure-api-keys-using-nextjs-server-components.mdx

## How to Secure API Keys Using Next.js Server Components

### Overview

> We can use [NextJS server components](https://nextjs.org/docs/app/building-your-application/rendering/server-components) to properly secure our API key from exposure in the front-end of our dapp. To further increase our API key security, we can also [restrict our API key to certain subgraphs or domains in Subgraph Studio](https://thegraph.com/docs/en/cookbook/upgrading-a-subgraph/#securing-your-api-key).

Reviewer: ```suggestion
We can use [NextJS server components](https://nextjs.org/docs/app/building-your-application/rendering/server-components) to properly secure our API key from exposure in the front-end of our dapp. To further increase our API key security, we can also [restrict our API key to certain subgraphs or domains in Subgraph Studio](/cookbook/upgrading-a-subgraph/#securing-your-api-key).
```
Prefer root-relative URLs (minus the locale) over absolute ones so navigating on staging or local dev doesn't direct you to production without you noticing. 😛

       <div>
           <h1>Server Component</h1>
           {data ? (
               <ul>
                   {data.factories.map((factory: any) => (
                       <li key={factory.id}>
                           <p>ID: {factory.id}</p>
                           <p>Pool Count: {factory.poolCount}</p>
                           <p>Transaction Count: {factory.txCount}</p>
                           <p>Total Volume USD: {factory.totalVolumeUSD}</p>
                       </li>
                   ))}
               </ul>
           ) : (
               <p>Loading data...</p>
           )}
       </div>
   );
}
```

### Step 4: Use the Server Component

1. In our page file (e.g., `pages/index.js`), import `ServerComponent`.
2. Render the component:

```javascript
import ServerComponent from './components/ServerComponent'

export default function Home() {
  return (
    <main>
      <ServerComponent />
    </main>
  )
}
```

### Step 5: Run and Test Our Dapp

Start our Next.js application using npm run dev. Verify that the server component is fetching data without exposing the API key.

![Server-side rendering](/img/api-key-server-side-rendering.png)

### Conclusion

> By utilizing Next.js Server Components, we've effectively hidden the API key from the client-side, enhancing the security of our application. This method ensures that sensitive operations are handled server-side, away from potential client-side vulnerabilities. Finally, be sure to explore [other API key security measures](https://thegraph.com/docs/en/cookbook/upgrading-a-subgraph/#securing-your-api-key) to increase your API key security even further.

Reviewer: ```suggestion
By utilizing Next.js Server Components, we've effectively hidden the API key from the client-side, enhancing the security of our application. This method ensures that sensitive operations are handled server-side, away from potential client-side vulnerabilities. Finally, be sure to explore [other API key security measures](/cookbook/upgrading-a-subgraph/#securing-your-api-key) to increase your API key security even further.
```

---
title: How to Secure API Keys Using Next.js Server Components
---

> ### Overview

Reviewer: ```suggestion
## Overview
```

Just to fix the lint error about the heading level...

I _think_ it's OK if "Caveats", "Why It's Needed", "Using client-side rendering to query a subgraph", and "Prerequisites" are under "Overview"?

Author: I think that's OK too.

> We can use [NextJS server components](https://nextjs.org/docs/app/building-your-application/rendering/server-components) to properly secure our API key from exposure in the front-end of our dapp. To further increase our API key security, we can also [restrict our API key to certain subgraphs or domains in Subgraph Studio](/cookbook/upgrading-a-subgraph/#securing-your-api-key).

Reviewer: @bermchain – Just noticed we sometimes write "frontend", sometimes "front-end". Does it matter?

Reviewer 2: @benface I think we can standardize as frontend

Reviewer: Thanks @bermchain, I'll standardize all the docs in another PR and will also check other codebases. Not sure how to quickly search in blog posts though, since they are in Notion and it looks like the database search only searches in properties, not page content... maybe not worth it for the blog?

> We can use [NextJS server components](https://nextjs.org/docs/app/building-your-application/rendering/server-components) to properly secure our API key from exposure in the front-end of our dapp. To further increase our API key security, we can also [restrict our API key to certain subgraphs or domains in Subgraph Studio](/cookbook/upgrading-a-subgraph/#securing-your-api-key).

Reviewer: ```suggestion
We can use [Next.js server components](https://nextjs.org/docs/app/building-your-application/rendering/server-components) to properly secure our API key from exposure in the front-end of our dapp. To further increase our API key security, we can also [restrict our API key to certain subgraphs or domains in Subgraph Studio](/cookbook/upgrading-a-subgraph/#securing-your-api-key).
```

> We can use [Next.js server components](https://nextjs.org/docs/app/building-your-application/rendering/server-components) to properly secure our API key from exposure in the front-end of our dapp. To further increase our API key security, we can also [restrict our API key to certain subgraphs or domains in Subgraph Studio](/cookbook/upgrading-a-subgraph/#securing-your-api-key).

Reviewer: ```suggestion
We can use [Next.js server components](https://nextjs.org/docs/app/building-your-application/rendering/server-components) to properly secure our API key from exposure in the frontend of our dapp. To further increase our API key security, we can also [restrict our API key to certain subgraphs or domains in Subgraph Studio](/cookbook/upgrading-a-subgraph/#securing-your-api-key).
```
(as recommended by Brian)

> In this cookbook, we will go over how to create a NextJS server component that queries a subgraph while also hiding the API key from the front-end.

Reviewer: ```suggestion
In this cookbook, we will go over how to create a Next.js server component that queries a subgraph while also hiding the API key from the front-end.
```

> In this cookbook, we will go over how to create a Next.js server component that queries a subgraph while also hiding the API key from the front-end.

Reviewer: ```suggestion
In this cookbook, we will go over how to create a Next.js server component that queries a subgraph while also hiding the API key from the frontend.
```

### Caveats

> - NextJS server components do not protect API keys from being drained using denial of service attacks.

Reviewer: ```suggestion
- Next.js server components do not protect API keys from being drained using denial of service attacks.
```

- The Graph Network gateways have denial of service detection and mitigation strategies in place, however using server components may weaken these protections.

> - NextJS server components introduce centralization risks as the server can go down.

Reviewer: ```suggestion
- Next.js server components introduce centralization risks as the server can go down.
```

### Why It's Needed

> In a standard React application, API keys included in the front-end code can be exposed to the client-side, posing a security risk. While `.env` files are commonly used, they don't fully protect the keys since React's code is executed on the client side, exposing the API key in the headers. Next.js Server Components address this issue by handling sensitive operations server-side.

Reviewer: ```suggestion
In a standard React application, API keys included in the frontend code can be exposed to the client-side, posing a security risk. While `.env` files are commonly used, they don't fully protect the keys since React's code is executed on the client side, exposing the API key in the headers. Next.js Server Components address this issue by handling sensitive operations server-side.
```

### Using client-side rendering to query a subgraph

![Client-side rendering](/img/api-key-client-side-rendering.png)

### Prerequisites

> - An API key from [www.thegraph.com/studio](www.thegraph.com/studio)

Reviewer: ```suggestion
- An API key from [Subgraph Studio](https://thegraph.com/studio)
```

- Basic knowledge of Next.js and React.

> - An existing Next.js project.

Reviewer 3: Many people can be using `pages/` and this won't be supported. So we should be explicit and can also link next docs https://nextjs.org/docs/app/building-your-application/rendering/server-components
```suggestion
- An existing Next.js project using app router
```

Author: How about this:

```suggestion
- An existing Next.js project that uses the [Next.js App Router](https://nextjs.org/docs/app).
```

> - An existing Next.js project that uses the [Next.js App Router](https://nextjs.org/docs/app).

Reviewer: ```suggestion
- An existing Next.js project that uses the [App Router](https://nextjs.org/docs/app).
```
Just a bit less repetitive.

## Step-by-Step Cookbook

### Step 1: Set Up Environment Variables

1. In our Next.js project root, create a `.env.local` file.
2. Add our API key: `API_KEY=<api_key_here>`.

### Step 2: Create a Server Component

1. In our `components` directory, create a new file, `ServerComponent.js`.
2. Use the provided example code to set up the server component.

### Step 3: Implement Server-Side API Request

In `ServerComponent.js`, add the following code:

> ```javascript

Reviewer: I'm surprised that Prettier indented these code snippets with 4 spaces instead of 2. 🤔

                       totalVolumeUSD
                   }
               }
           `,
           }),
       }
   );

   const responseData = await response.json();
   const data = responseData.data;

   return (
       <div>
           <h1>Server Component</h1>
           {data ? (
               <ul>
                   {data.factories.map((factory: any) => (
                       <li key={factory.id}>
                           <p>ID: {factory.id}</p>
                           <p>Pool Count: {factory.poolCount}</p>
                           <p>Transaction Count: {factory.txCount}</p>
                           <p>Total Volume USD: {factory.totalVolumeUSD}</p>
                       </li>
                   ))}
               </ul>
           ) : (
               <p>Loading data...</p>
           )}
       </div>
   );
}
```

### Step 4: Use the Server Component

1. In our page file (e.g., `pages/index.js`), import `ServerComponent`.
2. Render the component:

```javascript
import ServerComponent from './components/ServerComponent'

export default function Home() {
  return (
    <main>
      <ServerComponent />
    </main>
  )
}
```

### Step 5: Run and Test Our Dapp

> Start our Next.js application using npm run dev. Verify that the server component is fetching data without exposing the API key.

Reviewer: ```suggestion
Start our Next.js application using `npm run dev`. Verify that the server component is fetching data without exposing the API key.
```
